feat(init/meta/congr_tactic.lean): tactic to apply congruence rules

dselsam opened the pull request:
Features: 
1. a tactic `congr` to apply `eq`/`heq` congruence rules recursively.
2. a tactic `rel_congr` to apply the `rel_iff`/`rel_eq` rule.

Issues: I factored out the `rel_congr` tactic to avoid needing to try `rel_eq` on every application `congr`.

Commit:
feat(init/meta/congr_tactic.lean): tactic to apply congruence rules

gebner commented:
This PR has build errors, the test fails with:
```
-- testing congr_tactic.lean
LEAN ASSERTION VIOLATION
File: /home/travis/build/leanprover/lean/src/library/congr_lemma.cpp
Line: 596
is_app(a)
(C)ontinue, (A)bort/exit, (S)top/trap, Invoke (G)DB
failed congr_tactic.lean
```

leodemoura commented on library/init/meta/congr_tactic.lean:
```diff
@@ -0,0 +1,55 @@
+meta def apply_heq_congr_core (tgt : expr) : tactic unit :=
+do (lhs, rhs) ← (match_eq tgt <|> do (α, lhs, β, rhs) ← match_heq tgt, return (lhs, rhs)),
+   clemma ← mk_hcongr_lemma lhs^.get_app_fn lhs^.get_app_num_args,
+   apply_congr_core clemma
```
`^.` is deprecated. We use `.` now.

dselsam commented:
Fixed the assertion failure and removed the deprecated syntax. Haven't run the entire test-suite on Debug mode though, so still best to wait for the CI tests to run.
